### PR TITLE
Dockerizing tests

### PR DESCRIPTION
**Pull request by mvid**

### CODE REVIEW

**Comment by jackzampolin** [comment]
@alexanderbez

**Comment by alexanderbez** [comment]
@mvid mind giving a quick instruction set on how to test this? I can obviously start the Docker network via docker-compose, but how then do I test completeness? Are there commands or tests to run? Or is just starting the docker compose network sufficient?

**Comment by mvid** [comment]
@alexanderbez I am still working on completeness for the network, but right now I have a script in the root called `./start-testnet.sh` which should handle genesis, key signing, etc. After that, there will (probably) be another container that runs tests. Or maybe just a script.

**Comment by alexanderbez** [comment]
I see, so if I execute `./start-testnet.sh` with a zero exit-code, I can deem this PR acceptable?

File: ethereum/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+           --mine \
+           --http \
+           --http.port "8545" \
+           --http.addr "0.0.0.0" \
+           --http.corsdomain "*" \
+           --http.vhosts "*" \
+           --miner.threads=1 \
+           --nousb \
+           --verbosity=5 \
+           --miner.etherbase=0xBf660843528035a5A4921534E156a27e64B231fE
```
Comment (jackzampolin) [inline ethereum/Dockerfile:22]: This is a fun magic number.

File: solidity/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,18 @@
+FROM node:15.11-alpine3.13
+
+RUN apk update
+RUN apk add --no-cache python3 make g++ curl
+
+COPY . .
+RUN npm ci
+RUN chmod -R +x scripts
+
+RUN npm run typechain
```
Comment (jackzampolin) [inline solidity/Dockerfile:10]: Is there a way to do `npm` dependency caching here?

Comment (jackzampolin) [reply solidity/Dockerfile:10]: Related, has this been an issue as far as timing?

File: start-testnet.sh [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+echo "$n2name"_COSMOS_RPC="http://$n2name:26657" >> $n2dir/orchestrator.env
+echo "$n2name"_COSMOS_KEY=$(jq .priv_key.value $n2cfgDir/priv_validator_key.json) >> $n2dir/orchestrator.env
+echo "$n2name"_DENOM=stake >> $n2dir/orchestrator.env
+echo "$n2name"_ETH_RPC=http://ethereum:8545 >> $n2dir/orchestrator.env
+
+echo "$n3name"_COSMOS_GRPC="http://$n3name:9090" >> $n3dir/orchestrator.env
+echo "$n3name"_COSMOS_RPC="http://$n3name:26657" >> $n3dir/orchestrator.env
+echo "$n3name"_COSMOS_KEY=$(jq .priv_key.value $n3cfgDir/priv_validator_key.json) >> $n3dir/orchestrator.env
+echo "$n3name"_DENOM=stake >> $n3dir/orchestrator.env
+echo "$n3name"_ETH_RPC=http://ethereum:8545 >> $n3dir/orchestrator.env
```
Comment (jackzampolin) [inline start-testnet.sh:207]: Might be nice to break the above into a separate script where you could pass in a directory to populate as an arg.

File: solidity/scripts/contract-deployer.sh [hunk 4]
```diff
@@ -1,7 +1,7 @@
 #!/bin/bash
 npx ts-node \
 contract-deployer.ts \
---cosmos-node="http://localhost:26657" \
+--cosmos-node="http:///localhost:26657" \
```
Comment (jackzampolin) [inline solidity/scripts/contract-deployer.sh:4]: ?

File: ethereum/Dockerfile (hunk 1 above)
Comment (mvid) [reply ethereum/Dockerfile:22]: I got it from here: https://github.com/cosmos/gravity-bridge/blob/fd98d205b8db4575112b29735d77ecd569d67a1b/solidity/scripts/contract-deployer.sh#L6

If it doesn't have purpose, I will remove it. But it was there before

File: solidity/Dockerfile (hunk 2 above)
Comment (mvid) [reply solidity/Dockerfile:10]: It hasn't been an issue as far as timing. Also it should be cached by the docker build if nothing before it changed

File: start-testnet.sh (hunk 3 above)
Comment (mvid) [reply start-testnet.sh:207]: Yes, we should do that, and general prettyfication, after it works

File: solidity/scripts/contract-deployer.sh (hunk 4 above)
Comment (mvid) [reply solidity/scripts/contract-deployer.sh:4]: No idea. Removed.

File: solidity/Dockerfile (hunk 2 above)
Comment (mvid) [reply solidity/Dockerfile:10]: So far the slowest parts are deploying the contracts and building the orchestrator (rust build)

**Comment by zmanian** [comment]
To run `./start-testnet.sh` on my Mac I needed to 

`cd module && make install`
`brew install sponge`

`docker-compose logs orchestrator0` shows the timeout of the orachestor trying to communicate with `http://gravity0:1317`

A debug message shows the url is correct up to being passed to the Contact library

```
orchestrator0_1      | [peggy_utils/src/connection_prep.rs:109] &cosmos_legacy_url = "http://gravity0:1317"
```

**Comment by zmanian** [comment]
This is the only thing I've found in the dependencies for the orchestrator that might be an explanation 

https://github.com/actix/actix-web/issues/599

**Comment by zmanian** [comment]
@jkilpatr could you take a look? Does this meet your testing needs?

**Comment by zmanian** [comment]
@mvid can you fix the conflicts?

**Comment by mvid** [comment]
@zmanian conflicts fixed

**Comment by mvid** [comment]
Seeing this error in all but the first orchestrator:
`Valset cost estimate for Nonce 1 failed with Err(EthereumRestError(JsonRPCError { code: 3, message: "execution reverted: Supplied current validators and powers do not match checkpoint`

Any ideas @jkilpatr ?

**Comment by jackzampolin** [comment]
This looks like it is passing checks. Going to merge into the gentx branch

**Pull request merged**